### PR TITLE
Functions to generate parameters for Topmodel (TWI, and width function)

### PR DESCRIPTION
**Pull request by lcunha0118**
## Additions

Including functions to generate parameters for Topmodel (TWI, and width function)

## Testing

This code was tested based on the hydrofabrics release made in 11/05/2021. It runs standalone, independent of the topmodel BMI executable. It should be run before running topmodel, to generate the subcat.dat file. 

## Notes

-

## Todos

Test with hydrofabrics released in August/2021. 
Update README file

## Checklist

- [ ] PR has an informative and human-readable title
- [ ] Changes are limited to a single goal (no scope creep)
- [ ] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [ ] Passes all existing automated tests
- [ ] Any _change_ in functionality is tested
- [ ] New functions are documented (with a description, list of inputs, and expected output)
- [ ] Placeholder code is flagged / future todos are captured in comments
- [ ] Visually tested in supported browsers and devices (see checklist below :point_down:)
- [ ] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [ ] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

### Target Environment support

- [ ] Windows
- [ x] Linux
- [ ] Browser

### Accessibility

- [ ] Keyboard friendly
- [ ] Screen reader friendly

### Other

- [ ] Is useable without CSS
- [ ] Is useable without JS
- [ ] Flexible from small to large screens
- [ ] No linting errors or warnings
- [ ] JavaScript tests are passing

### CODE REVIEW

File: params/README.md [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ 	- osgeo (gdal,ogr)
+ 	- numpy
+ 	- agparse
+ 	- pandas
+ 	
+3) Curl to download the HAND DEM data
+
+## Usage
+1) Edit the workflow_hand_twi_giuh.env file. This file contains all parameters for the runs, including the HUC06 number, path to the hydrofabrics, and environmental variabels for TAUDEM and gdal
+2) Run: source workflow_hand_twi_giuh.sh
```
Comment (SnowHydrology) [inline params/README.md:34]: In the future, we can add code snippets or blocks when instructing the user to run something from the command line

File: params/others/Run_topmodel.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+sys.path.append(BASE_DIR+"/params/src/")
+import summarize_results_functions as FC
+
+outputfolder_results=BASE_DIR+"/params/data/TopModel_Results/"
+if not os.path.exists(outputfolder_results): os.mkdir(outputfolder_results)
+
+run_flag=1
+process_results_flag=1
+Resolution=30    
+#hydro_fabrics_input_dir="/home/west/Projects/hydrofabrics/20210511/"
```
Comment (SnowHydrology) [inline params/others/Run_topmodel.py:27]: OWP guidelines say we cannot include paths that reveal a computer's directory structure. Please remove.

File: params/others/Run_topmodel.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+#outputfolder_twi="/home/west/Projects/hydrofabrics/20210511/TWI_"+str(Resolution)+"m/TOPMODEL_cat_file/"
```
Comment (SnowHydrology) [inline params/others/Run_topmodel.py:28]: Please remove.

File: params/others/summarize_results_V2.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+Resolution=30
+method=1
+input_twi=BASE_DIR+"/params/data/hydrofabrics/releases/beta/01a/TWI_30m/TOPMODEL_cat_file/"
+outputfolder_summary=BASE_DIR+"/params/data/plots/"
+if not os.path.exists(outputfolder_summary): os.mkdir(outputfolder_summary)
+
+input_giuh=BASE_DIR+"/params/data/hydrofabrics/releases/beta/01a/GIUH_30m_1/CFE_config_file/"
+outputfolder_summary=BASE_DIR+"/params/data/plots/"
+if not os.path.exists(outputfolder_summary): os.mkdir(outputfolder_summary)
+# FOR CFE - GIUUH outputfolder_giuh="/home/west/Projects/hydrofabrics/20210511/GIUH_"+str(Resolution)+"m_"+str(method)+"/CFE_config_file/"
```
Comment (SnowHydrology) [inline params/others/summarize_results_V2.py:35]: Please remove

File: params/others/summarize_results_V2.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
+# -*- coding: utf-8 -*-
```
Comment (SnowHydrology) [inline params/others/summarize_results_V2.py:1]: Please change file name to summarize_results.py

File: params/src/generate_giuh_per_basin.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+    y1 = int((bbox[3] - originY) / pixel_height)
+    y2 = int((bbox[2] - originY) / pixel_height) 
+    #y2 = int((bbox[2] - originY) / pixel_height) + 1
+     
+    xsize = x2 - x1
+    ysize = y2 - y1
+    return (x1, y1, xsize, ysize)
+
+# namest='020302'
+# catchments='/home/west/Projects/hydrofabrics/20210511/catchments_wgs84.geojson'
```
Comment (SnowHydrology) [inline params/src/generate_giuh_per_basin.py:54]: Please remove

File: params/src/generate_giuh_per_basin.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+                    nodata_value = args.nodata,
+                    global_src_extent = args.preload,
+                    buffer_distance = args.buffer,
+                    output_flag = args.output,                    
+                    )
+
+
+# Test
+namest='010100'
+catchments='/home/west/Projects/hydrofabrics/20210511/catchments_wgs84.geojson'
```
Comment (SnowHydrology) [inline params/src/generate_giuh_per_basin.py:319]: Please remove

**Review by SnowHydrology — COMMENTED** [review]
Thanks for putting this together @lcunha0118! I only have some minor comments. The main recurring one would be to remove all references to the west computer path (I found a few, but please do a more detailed search). The python scripts could also do with a bit more commenting. We can worry about that later, though.

**Comment by madMatchstick** [comment]
Add any results driven files or folders to .gitignore

**Review by SnowHydrology — APPROVED** [review]
Further testing needed on other machines because of GDAL and TauDEM versioning.

**Pull request merged**